### PR TITLE
[USER32][WIN32SS] Improve CascadeWindows function

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
This is an improvement of CascadeWindows in user32.
JIRA issue: [CORE-14807](https://jira.reactos.org/browse/CORE-14807)

## Proposed changes
- If the window is too large, then shrink it if possible.

### CODE REVIEW

File: win32ss/user/user32/windows/mdi.c [hunk 1]
```diff
@@ -2047,6 +2047,27 @@ GetCascadeChildProc(HWND hwnd, LPARAM lParam)
...
+static void
+QuerySizeFix(HWND hwnd, LPINT pcx, LPINT pcy)
+{
+    MINMAXINFO mmi;
+    DWORD dwResult;
+
+    mmi.ptMinTrackSize.x = mmi.ptMinTrackSize.y = 0;
+    mmi.ptMaxTrackSize.x = mmi.ptMaxTrackSize.y = MAXLONG;
+    SendMessageTimeoutW(hwnd, WM_GETMINMAXINFO, 0, (LPARAM)&mmi,
+                        SMTO_ABORTIFHUNG, 80, &dwResult);
```
Comment (learn-more) [inline win32ss/user/user32/windows/mdi.c]: How should we handle a window that is hanging?
On windows, you can test this by adding a `Sleep(40 * 1000)` in the `WM_GETMINMAXINFO` handler. (Or any other timeout)

File: win32ss/user/user32/windows/mdi.c [hunk 2]
```diff
@@ -2112,30 +2135,51 @@ CascadeWindows(HWND hwndParent, UINT wFlags, LPCRECT lpRect,
...
-        cx = rcWnd.right - rcWnd.left;
-        cy = rcWnd.bottom - rcWnd.top;
+        cxNew = cx = rcWnd.right - rcWnd.left;
+        cyNew = cy = rcWnd.bottom - rcWnd.top;
+
+        /* if we can change the window size */
+        if (GetWindowLong(hwnd, GWL_STYLE) & WS_THICKFRAME)
+        {
+            /* check the size */
+#define THRESHOLD(xy) (((xy) * 5) / 7)      /* in the rate 5/7 */
```
Comment (learn-more) [inline win32ss/user/user32/windows/mdi.c]: Is this also the same threshold if there is only 1 window?

File: win32ss/user/user32/windows/mdi.c (hunk 1 above)
Comment (katahiromz) [reply win32ss/user/user32/windows/mdi.c]: I tried with WM_COMMAND. If timeout has come, SendMessageTimeout returns zero.

File: win32ss/user/user32/windows/mdi.c (hunk 2 above)
Comment (katahiromz) [reply win32ss/user/user32/windows/mdi.c]: Special handling is added.

File: win32ss/user/user32/windows/mdi.c [hunk 3]
```diff
@@ -2112,30 +2115,63 @@ CascadeWindows(HWND hwndParent, UINT wFlags, LPCRECT lpRect,
...
+
+        /* if we can change the window size */
+        if ((GetWindowLong(hwnd, GWL_STYLE) & WS_THICKFRAME) == WS_THICKFRAME)
+        {
+            /* check the size */
+#define THRESHOLD(xy) (((xy) * 5) / 7)      /* in the rate 5/7 */
+            if (cxNew > THRESHOLD(cxWork))
+                cxNew = THRESHOLD(cxWork);
+            if (cyNew > THRESHOLD(cyWork))
+                cyNew = THRESHOLD(cyWork);
```
Comment (HBelusca) [inline win32ss/user/user32/windows/mdi.c]:
```c
cxNew = min(cxNew, THRESHOLD(cxWork));
cyNew = min(cyNew, THRESHOLD(cyWork));
```

File: win32ss/user/user32/windows/mdi.c [hunk 4]
```diff
@@ -2112,30 +2136,51 @@ CascadeWindows(HWND hwndParent, UINT wFlags, LPCRECT lpRect,
...
             ShowWindow(hwnd, SW_RESTORE | SW_SHOWNA);
 
         GetWindowRect(hwnd, &rcWnd);
-        cx = rcWnd.right - rcWnd.left;
-        cy = rcWnd.bottom - rcWnd.top;
+        cxNew = cx = rcWnd.right - rcWnd.left;
+        cyNew = cy = rcWnd.bottom - rcWnd.top;
+
+        /* if we can change the window size and it is not only one */
+        if (info.chwnd != 1 && (GetWindowLong(hwnd, GWL_STYLE) & WS_THICKFRAME))
```
Comment (HBelusca) [inline win32ss/user/user32/windows/mdi.c]: Please use GetWindowLongPtrW.

File: win32ss/user/user32/windows/mdi.c [hunk 5]
```diff
@@ -2047,6 +2047,28 @@ GetCascadeChildProc(HWND hwnd, LPARAM lParam)
...
+                            SMTO_ABORTIFHUNG | SMTO_NORMAL, 120, &dwResult))
+    {
+        if (*pcx < mmi.ptMinTrackSize.x)
+            *pcx = mmi.ptMinTrackSize.x;
+        if (*pcy < mmi.ptMinTrackSize.y)
+            *pcy = mmi.ptMinTrackSize.y;
+        if (*pcx > mmi.ptMaxTrackSize.x)
+            *pcx = mmi.ptMaxTrackSize.x;
+        if (*pcy > mmi.ptMaxTrackSize.y)
+            *pcy = mmi.ptMaxTrackSize.y;
```
Comment (HBelusca) [inline win32ss/user/user32/windows/mdi.c]: You can use the min/max macros to simplify the code:
```c
*pcx = min(max(*pcx, mmi.ptMinTrackSize.x), mmi.ptMaxTrackSize.x);
*pcy = min(max(*pcy, mmi.ptMinTrackSize.y), mmi.ptMaxTrackSize.y);
```

File: win32ss/user/user32/windows/mdi.c (hunk 4 above)
Comment (katahiromz) [reply win32ss/user/user32/windows/mdi.c]: Usage of GetWindowLong for GWL_STYLE is correct even in Win64. There is no "GWLP_STYLE".

File: win32ss/user/user32/windows/mdi.c (hunk 3 above)
Comment (katahiromz) [reply win32ss/user/user32/windows/mdi.c]: OK.

File: win32ss/user/user32/windows/mdi.c (hunk 4 above)
Comment (katahiromz) [reply win32ss/user/user32/windows/mdi.c]: Used GetWindowLongW.

File: win32ss/user/user32/windows/mdi.c (hunk 5 above)
Comment (katahiromz) [reply win32ss/user/user32/windows/mdi.c]: Ok.
■

File: win32ss/user/user32/windows/mdi.c (hunk 4 above)
Comment (katahiromz) [reply win32ss/user/user32/windows/mdi.c]: Used GetWindowLongPtrW.

**Pull request merged**